src/main/java/org/jboss/threads/EnhancedQueueExecutor.java

```diff
@@ -1515,16 +1522,33 @@ private QNode getOrAddNode(PoolThreadNode nextPoolThreadNode) {
             for (;;) {
                 head = EnhancedQueueExecutor.this.head;
                 headNext = head.getNext();
+                // this happen if another consumer has already consumed head:
+                // retry with a fresh head
+                if (headNext == head) {
+                    if (UPDATE_STATISTICS) spinMisses.increment();
+                    JDKSpecific.onSpinWait();
+                    continue;
+                }
                 if (headNext instanceof TaskNode) {
                     TaskNode taskNode = (TaskNode) headNext;
                     if (compareAndSetHead(head, taskNode)) {
+                        // save from GC Nepotism: generational GCs don't like
+                        // cross-generational references, so better to "clean-up" head::next
+                        // to save dragging head::next into the old generation.
+                        // Clean-up cannot just null out next
+                        head.setNextOrdered(head);
```

Comment:
I need to see a correctness proof here.  What are the possible observable output states, and why?  Are the states all valid and already described above?  What are the preconditions and postconditions?

Reply:
It's mostly expressed in term of acquire-release relationships between atomic operations: 

- only one thread can win compareAndSetHead(head, taskNode) for a specific head, taskNode pair
- other threads can reach head while its head::next is valued differently:

1. "slow" producers (ie producers descheduled after reading this.tail for enough time that other producers have moved on and a consumer has already moved head): 
- before `head.setNextOrdered(head)`: they will try to `compareAndSetNext(null, newValue)` and will fail, because given that the consumer has been able to move past to it, it means that next != null
- after `head.setNextOrdered(head)`: they will ry to `compareAndSetNext(null, newValue)` and will fail, because head != null
2. slow consumers (ie consumers that have been descheduled after reading this.head and are going to read head::getNext)
- before `head.setNextOrdered(head)`: they will get the same value of the consumer that has won casHead, hence they will attempts casHead, failing because has been already moved on
- after `head.setNextOrdered(head)`: they will get the new value (head::next == head) and they will reattempt reading this.head. Given that head.setNextOrdered is write releasing any previous write (including the won casHead), load head::next and finding == head is load acquiring the new value of head, hence reading head will return *at least* the value released by the consumer that has won casHead

Reply:
The states already written in the overview of the class should be valid, because I've tried hard to not break the way we use `null` and just add checks that head::next == next means that "something" has changed: in particular `head` and by consequence `tail`.

Reply:
OK I think I understand all the possibilities here. But couldn't this be `setNextRelaxed`? We're not using `node.next == node` as the solitary indicator that it has been removed, so it's probably OK if some threads don't observe that (they would just spin past the node using the current behavior).  Does it make a difference performance-wise?

Reply:
> Does it make a difference performance-wise?
setNextOrdered on x86 isn't using any specific hardware barrier; thanks to the strong memory model of x86 and the only perf hit would have been by using "updaters" but yet negligible.

From a correctness pov yes: the only thing we care about here is *atomicity*, but sadly `Unsafe` isn't providing the right methods to just have that: ideally this is what we need here https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/VarHandle.AccessMode.html#SET_OPAQUE , because the compareAndSet previous to it already provide the right sequential consistent write we need.

Reply:
Ah, I see.

Reply:
Is this safe in cases where `EQE.head` passes `EQE.tail`?

The tail reference may lag (currently by only one node, but the potential jdk14 CQL optimization could potentially allow more lag), which I think allows `EQE.head` to be `EQE.tail.next`. Clearing references may unlink tail from head and orphan submitted tasks.

It's very possible that I've missed an important piece here, I'm not confident that I've gotten this right.

Reply:
It shouldn't AFAIK.
A producer set prev_tail::casNext(null, new_tail) and then casTail(prev_tail,new_tail), while a consumer can consume only prev_tail, not new_tail (yet).
Given that other producers are able to read `this.tail` (ie new_tail) and evetually `this.tail == new_tail` others are always able to append new element without loosing bits around

Reply:
@carterkozak Thinking about it again and again, probably we're "lucky" that we cannot lag, but if we would start to use the jdk 14 optimization (as you've suggested) we need, on 
 producer side, chase the last tail starting from head (that in theory should be near(ish) given that has past tail, that has lagged behind, if node.next == node (see https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/concurrent/ConcurrentLinkedQueue.java#L348).

Reply:
> From a correctness pov yes: the only thing we care about here is atomicity
@franz1981 you can use a plain write in this case, since reference writes are always atomic

Reply:
@nitsanw thanks, yep JLS should ensure that things ref equality should work fine here, no tearing on the reference value indeed, thanks!